src/main/java/com/minecolonies/structures/helpers/Structure.java

```diff
@@ -84,61 +97,340 @@
      */
     public Structure(@Nullable final World world, final String structureName, final PlacementSettings settings)
     {
-        InputStream inputstream = MinecraftServer.class.getResourceAsStream("/assets/" + Constants.MOD_ID + "/schematics/" + structureName + ".nbt");
-
+        String correctStructureName = structureName;
         if (world == null || world.isRemote)
         {
             this.settings = settings;
             this.mc = Minecraft.getMinecraft();
         }
 
-        //Might be at a different location!
-        if (inputstream == null)
+        InputStream inputStream = null;
+        //Try the cache first
+        if (Structures.hasMD5(correctStructureName))
         {
-            try
+            inputStream = Structure.getStream(Structures.SCHEMATICS_CACHE + '/' + Structures.getMD5(correctStructureName));
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png 'Severity: BLOCKER') Close this "InputStream". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2095)